### DISCUSSION TITLE
Significance of the create_subscription parameter

### DISCUSSION THREAD

**Discussion by dibyendumca** [FreeOpcUa/opcua-asyncio#1151] [Q&A]
Hello,
Can anybody explain the significance of the 1st parameter of the client function 'CREATE_SUBSCRIPTION' ?

```
handler = SubscriptionHandler()
client.create_subscription(500, handler)
```

**_What this value 500 means exactly?_**
**_What will happen if we increase or decrease this value?_**

**Comment by schroeder-** [comment]
The first parameter is **publishing interval**. For a complete description see [opc ua ref](https://reference.opcfoundation.org/v104/Core/docs/Part4/5.12.1/). 
A simple explanation is that is the interval in ms(milliseconds), in which the server checks if a monitored item has changed, in your subscription. Lowering only works if the samplingrate of a VariableNode is lower than the publishing interval.

**Reply by dibyendumca** [reply]
Thank you very much

**Reply by dibyendumca** [reply]
I am getting the below line in the console as and when I am running a Server module of an ASYNCUA.

```
Subscription Subscription(id:78) has expired, publish cycle count(10001) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(10002) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(10003) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(10003) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(10004) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(10005) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(10006) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(10007) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(10008) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(10009) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(10010) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(10011) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(10012) > lifetime count (10000)
......
Subscription Subscription(id:78) has expired, publish cycle count(66106) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(66107) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(66107) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(66108) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(66109) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(66110) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(66110) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(66111) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(66112) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(66112) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(66113) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(66114) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(66115) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(66116) > lifetime count (10000)
Subscription Subscription(id:78) has expired, publish cycle count(66117) > lifetime count (10000)
```

May I know what is this about?

**Reply by AndreasHeine** [reply]
> I am getting the below line in the console as and when I am running a Server module of an ASYNCUA.
> 
> ```
> Subscription Subscription(id:78) has expired, publish cycle count(10001) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(10002) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(10003) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(10003) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(10004) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(10005) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(10006) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(10007) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(10008) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(10009) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(10010) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(10011) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(10012) > lifetime count (10000)
> ......
> Subscription Subscription(id:78) has expired, publish cycle count(66106) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(66107) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(66107) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(66108) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(66109) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(66110) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(66110) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(66111) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(66112) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(66112) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(66113) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(66114) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(66115) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(66116) > lifetime count (10000)
> Subscription Subscription(id:78) has expired, publish cycle count(66117) > lifetime count (10000)
> ```
> 
> May I know what is this about?

https://reference.opcfoundation.org/v104/Core/docs/Part4/5.13.1/

[Subscriptions](https://reference.opcfoundation.org/search/17?t=Subscriptions)have a lifetime counter that counts the number of consecutive publishing cycles in which there have been no [Publish](https://reference.opcfoundation.org/search/17?t=Publish)requests available to send a [Publish](https://reference.opcfoundation.org/search/17?t=Publish)response for the [Subscription](https://reference.opcfoundation.org/search/17?t=Subscription). Any [Service](https://reference.opcfoundation.org/search/17?t=Service)call that uses the [SubscriptionId](https://reference.opcfoundation.org/search/17?t=SubscriptionId)or the processing of a [Publish](https://reference.opcfoundation.org/search/17?t=Publish)response resets the lifetime counter of this [Subscription](https://reference.opcfoundation.org/search/17?t=Subscription). When this counter reaches the value calculated for the lifetime of a [Subscription](https://reference.opcfoundation.org/search/17?t=Subscription)based on the MaxKeepAliveCount parameter in the [CreateSubscription](https://reference.opcfoundation.org/search/17?t=CreateSubscription) [Service](https://reference.opcfoundation.org/search/17?t=Service)([5.13.2](https://reference.opcfoundation.org/Core/Part4/v104/docs/?r=_Ref113157332)), the [Subscription](https://reference.opcfoundation.org/search/17?t=Subscription)is closed. Closing the [Subscription](https://reference.opcfoundation.org/search/17?t=Subscription)causes its [MonitoredItems](https://reference.opcfoundation.org/search/17?t=MonitoredItems)to be deleted. In addition the [Server](https://reference.opcfoundation.org/search/17?t=Server)shall issue a [StatusChangeNotification](https://reference.opcfoundation.org/search/17?t=StatusChangeNotification) [notificationMessage](https://reference.opcfoundation.org/search/17?t=notificationMessage)with the status code Bad_Timeout. The [StatusChangeNotification notificationMessage](https://reference.opcfoundation.org/search/17?t=StatusChangeNotification%20notificationMessage)type is defined in [7.20.4](https://reference.opcfoundation.org/Core/Part4/v104/docs/?r=_Ref166078476).

**Reply by AndreasHeine** [reply]
it basically means that the client does not send any publishrequest for a while!
Client -> PublishRequest -> Server -> Subscription
if no data:
Subscription -> Server -> KeepAlive -> Client
then the client needs to send a new PublishRequest!
but in your case this does not happen with "Subscription(id:78)"!

edit:

it often happens if the publishinterval is to small e.g. <500ms and you have "static" data like configuration parameters which do not change frequently... so you will get many keepalives...

**Reply by JamesTev** [reply]
@AndreasHeine I am having a similar problem to the one mentioned above. Your note "it often happens if the publishinterval is to small" seems like it might be the cause for me as I have some variables that may not change for many hours, despite my subscription period being 500ms. What do you suggest I do to get around this? And how can I catch subscriptions being deleted?

**Reply by AndreasHeine** [reply]
you can add the servertime variable to the subscription! wich would be the easiest fix.

https://www.linkedin.com/pulse/brief-introduction-opc-ua-subscriptions-andreas-heine-3pc6e?utm_source=share&utm_medium=member_android&utm_campaign=share_via
